src/library_pthread.js

```diff
@@ -1069,6 +1086,111 @@ var LibraryPThread = {
+      func.apply(null, args);
+    });
+  },
+
+  _emscripten_promise_resolve__deps: ['$promiseMap'],
+  _emscripten_promise_resolve__sig: 'vip',
+  _emscripten_promise_resolve: function(id, value) {
+#if RUNTIME_DEBUG
+    err('emscripten_resolve_promise: ' + id);
+#endif
+    assert(promiseMap.has(id));
+    promiseMap.get(id).resolve(value);
+    promiseMap.delete(id);
+  },
+
+  _emscripten_promise_reject__deps: ['$promiseMap'],
+  _emscripten_promise_reject__sig: 'vip',
+  _emscripten_promise_reject: function(id) {
+#if RUNTIME_DEBUG
+    dbg('emscripten_promise_reject: ' + id);
+#endif
+    assert(promiseMap.has(id));
+    promiseMap.get(id).reject();
+    promiseMap.delete(id);
+  },
+
+  // Asynchronous version sync_all_threads. Always run on the main thread.
+  // This work happens asynchronously. The `callback` is called once this work
+  // is completed, passing the ctx.
+  // TODO(sbc): Should we make a new form of __proxy attribute for JS library
+  // function that run asynchronously like but blocks the caller until they are
+  // done.  Perhaps "sync_with_ctx"?
+  _emscripten_sync_all_threads_async__sig: 'viii',
+  _emscripten_sync_all_threads_async__deps: ['_emscripten_proxy_sync_code_async', '$newNativePromise'],
+  _emscripten_sync_all_threads_async: function(caller, callback, ctx) {
+#if PTHREADS_DEBUG
+    dbg("_emscripten_sync_all_threads_async caller=" + ptrToString(caller));
+#endif
+#if ASSERTIONS
+    assert(!ENVIRONMENT_IS_PTHREAD, 'Internal Error! _emscripten_sync_all_threads_async() can only ever be called from main thread');
+#endif
+
+    let promises = [];
+
+    // This first promise resolves once the main thread has loaded all modules.
+    promises.push(newNativePromise(__emscripten_thread_sync_code_async, []));
+
+    // We then create a sequence of promises, one per thread, that resolve once
+    // each thread has performed its sync using _emscripten_proxy_sync_code.
+    // Any new threads that are created after this call will automaticaly be
+    // in sync because we call `__emscripten_thread_sync_code` in
+    // invokeEntryPoint before the threads entry point is called.
+    for (const ptr of Object.keys(PThread.pthreads)) {
+      const pthread_ptr = Number(ptr);
+      if (pthread_ptr !== caller && !(pthread_ptr in PThread.zombieThreads)) {
+        promises.push(newNativePromise(__emscripten_proxy_sync_code_async, [pthread_ptr]));
+      }
+    }
+
+#if PTHREADS_DEBUG
+    dbg('_emscripten_sync_all_threads_async: waiting on ' + promises.length + ' promises');
+#endif
+    // Once all promises are resolved then we know all threads are in sync and
+    // we can call the callback.
+    Promise.all(promises).then(() => {
+#if PTHREADS_DEBUG
+      dbg('_emscripten_sync_all_threads_async done: calling callback');
+#endif
+      {{{ makeDynCall('vp', 'callback') }}}(ctx);
+    });
+  },
+#endif // MAIN_MODULE
+
+  // Asynchronous version sync_all_threads. Always run on the main thread.
```

Comment:
I'm finding this (and the bit that calls it below) quite confusing. The comment says it's async (and the assertion) but the version above is actually named _async.

Reply:
Sorry, typo here in this comment.  Fixed.

Reply:
Still working on how the name these things.. will finalize that once all the tests pass